File: main/src/main/java/com/sedmelluq/discord/lavaplayer/source/soundcloud/SoundCloudHelper.java
package com.sedmelluq.discord.lavaplayer.source.soundcloud;

import com.sedmelluq.discord.lavaplayer.tools.JsonBrowser;
import com.sedmelluq.discord.lavaplayer.tools.io.HttpClientTools;
import com.sedmelluq.discord.lavaplayer.tools.io.HttpInterface;
import com.sedmelluq.discord.lavaplayer.tools.io.PersistentHttpStream;
import java.io.IOException;
import java.net.URI;

public class SoundCloudHelper {
  public static String nonMobileUrl(String url) {
    if (url.startsWith("https://m.")) {
      return "https://" + url.substring("https://m.".length());
    } else {
      return url;
    }
  }

  public static String loadPlaybackUrl(HttpInterface httpInterface, String jsonUrl) throws IOException {
    try (PersistentHttpStream stream = new PersistentHttpStream(httpInterface, URI.create(jsonUrl), null)) {
      if (!HttpClientTools.isSuccessWithContent(stream.checkStatusCode())) {
        throw new IOException("Invalid status code for soundcloud stream: " + stream.checkStatusCode());
      }

      JsonBrowser json = JsonBrowser.parse(stream);
      return json.get("url").text();
    }
  }
}

File: main/src/main/java/com/sedmelluq/discord/lavaplayer/source/soundcloud/SoundCloudHelper.kt
package com.sedmelluq.discord.lavaplayer.source.soundcloud

import com.sedmelluq.discord.lavaplayer.tools.io.HttpClientTools
import com.sedmelluq.discord.lavaplayer.tools.io.HttpInterface
import com.sedmelluq.discord.lavaplayer.tools.io.PersistentHttpStream
import com.sedmelluq.discord.lavaplayer.tools.json.JsonBrowser
import java.io.IOException
import java.net.URI

object SoundCloudHelper {
    @JvmStatic
    fun nonMobileUrl(url: String): String {
        return url.takeIf { !url.startsWith("https://m.") }
            ?: "https://${url.substring("https://m.".length)}"
    }

    @Throws(IOException::class)
    @JvmStatic
    fun loadPlaybackUrl(httpInterface: HttpInterface, jsonUrl: String): String {
        PersistentHttpStream(httpInterface, URI.create(jsonUrl), null).use { stream ->
            if (!HttpClientTools.isSuccessWithContent(stream.checkStatusCode())) {
                throw IOException("Invalid status code for soundcloud stream: " + stream.checkStatusCode())
            }

            return JsonBrowser.parse(stream)["url"].safeText
        }
    }
}
